### PR TITLE
MMA-10597 new all replicas endpoint 

### PR DESCRIPTION
**Pull request by thomaskwscott**
This PR adds a new ListAllReplicasAction that adds the following endpoint:

/v3/clusters/{clusterId}/topics/-/replicas

This returns the replicas for all topics and all partitions.

Tested with new unit and integration tests.

### CODE REVIEW

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ReplicasResource.java [hunk 1]
```diff
@@ -114,12 +116,17 @@ public void getReplica(
...
+            .thenApply(replica -> GetReplicaResponse.create(toReplicaData(
+                replica,
+                crnFactory,
+                urlFactory)));
 
     AsyncResponses.asyncResume(asyncResponse, response);
   }
 
-  private ReplicaData toReplicaData(PartitionReplica replica) {
+  public static ReplicaData toReplicaData(PartitionReplica replica,
```
Comment (srpanwar-confluent) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ReplicasResource.java:127]: @thomaskwscott looks like `TopicPartitionInfo.observer` information is not available in the Confluent Community version of Kafka lib. We show `Observer` info on the Topic overview page. 
We can either merge this implementation as is and add a new endpoint in ce-kafka-extension to fetch observer info or move this to change to `ce-kafka-extension` as is. Either way is fine with me.

Comment (thomaskwscott) [reply kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ReplicasResource.java:127]: oh, I didn't realise we were after observer info. I'll move this to ce-kafka.

Comment (thomaskwscott) [reply kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ReplicasResource.java:127]: @srpanwar-confluent just to confirm is it the /replicas or /replica-status endpoint we need a cluster level version of?

File: api/v3/openapi.yaml [hunk 2]
```diff
@@ -624,6 +624,19 @@ paths:
         '200':
           $ref: '#/components/responses/ListReplicasResponse'
 
+  /clusters/{cluster_id}/topics/-/replicas:
```
Comment (rigelbm) [inline api/v3/openapi.yaml:627]: This should be `/clusters/{cluster_id}/topics/-/partitions/-/replicas`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ReplicaData.java [hunk 3]
```diff
@@ -61,6 +61,23 @@ public static Builder fromPartitionReplica(PartitionReplica replica) {
         .setInSync(replica.isInSync());
   }
 
+  public static ReplicaData create(
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ReplicaData.java:64]: Why not use the `Builder` available?

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ListAllReplicasAction.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.MediaType;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+@Path("/v3/clusters/{clusterId}/topics/-/replicas")
+@ResourceName("api.v3.topic-configs.*")
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ListAllReplicasAction.java:43]: This resource name should be `api.v3.replicas.*`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ListAllReplicasAction.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      Provider<TopicManager> topicManager, CrnFactory crnFactory, UrlFactory urlFactory) {
+    this.topicManager = requireNonNull(topicManager);
+    this.crnFactory = requireNonNull(crnFactory);
+    this.urlFactory = requireNonNull(urlFactory);
+  }
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @PerformanceMetric("v3.replicas.list")
+  @ResourceName("api.v3.replicas.list")
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ListAllReplicasAction.java:61]: This should be `api.v3.replicas.list-all`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ListAllReplicasAction.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  public ListAllReplicasAction(
+      Provider<TopicManager> topicManager, CrnFactory crnFactory, UrlFactory urlFactory) {
+    this.topicManager = requireNonNull(topicManager);
+    this.crnFactory = requireNonNull(crnFactory);
+    this.urlFactory = requireNonNull(urlFactory);
+  }
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @PerformanceMetric("v3.replicas.list")
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ListAllReplicasAction.java:60]: `v3.replicas.list-all`

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReplicasActionIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class ListAllReplicasActionIntegrationTest extends ClusterTestHarness {
+
+    private static final String TOPIC_1 = "topic-1";
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReplicasActionIntegrationTest.java:36]: The formatting on this file looks off. We use 2-space ident, 4-space continuation.

File: kafka-rest/src/test/java/io/confluent/kafkarest/resources/v3/ListAllReplicasActionTest.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+  private static final PartitionReplica REPLICA_1 =
+      PartitionReplica.create(
+          CLUSTER_ID,
+          TOPIC_NAME,
+          0,
+          1,
+          /* isLeader= */ true,
+          /* isInSync= */ true);
+  private static final PartitionReplica REPLICA_2 =
+          PartitionReplica.create(
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/resources/v3/ListAllReplicasActionTest.java:59]: Formatting becomes off starting here as well.

**Comment by thomaskwscott** [comment]
after discussing with @srpanwar-confluent this should be in ce-kafka-rest and should be replica-status instead. Will close this one and open a new PR for it.

**Pull request closed without merging**